src/software/thunderscope/play/qlearninginfo_widget.py

```diff
@@ -0,0 +1,242 @@
+
+        num_features = linear_q_func_info.num_features
+        num_actions = linear_q_func_info.num_actions
+
+        # Setting table size dynamically
+        self.linear_q_func_weights_table.setColumnCount(num_features)
+        self.linear_q_func_weights_table.setRowCount(num_actions)
+
+        # Set table data with weights
+        for row in range(num_actions):
+            for col in range(num_features):
+                weight = str(linear_q_func_info.weights[row * num_features + col])
+                table_item = QTableWidgetItem(weight)
+                table_item.setSizeHint(
+                    QtCore.QSize(
+                        len(weight)
+                        * QLearningInfoWidget.ITEM_SIZE_HINT_WIDTH_EXPANSION,
+                        1,
+                    )
+                )
+                table_item.setFlags(
+                    table_item.flags() & ~QtCore.Qt.ItemFlag.ItemIsEditable
+                )
+                self.linear_q_func_weights_table.setItem(row, col, table_item)
+
+        # Set row and column headers
+        self.linear_q_func_weights_table.setHorizontalHeaderLabels(
+            linear_q_func_info.feature_names
+        )
+        self.linear_q_func_weights_table.setVerticalHeaderLabels(
+            linear_q_func_info.action_names
+        )
+
+    def __display_action_selection_strategy_info(self, mdp_name: str) -> None:
+        """Update the BarGraphWidget with data from the most recent 
+        ActionSelectionStrategyInfo for the given MDP
+        
+        :param mdp_name: the name identifying the MDP
+        """
+        # Check if we have an ActionSelectionStrategyInfo for the given MDP
+        if (mdp_name not in self.mdp_info_dict) or (
+            not self.mdp_info_dict[mdp_name].action_selection_strategy_info
+        ):
+            return
+
+        action_selection_strategy_info = self.mdp_info_dict[
+            mdp_name
+        ].action_selection_strategy_info
+
+        bar_graph_data = {}
+        bar_graph_colours = {}
+        for action in action_selection_strategy_info.actions:
+            bar_graph_data[action.name] = action.value
+            if action.selected:
+                bar_graph_colours[
+                    action.name
+                ] = QLearningInfoWidget.SELECTED_ACTION_BAR_COLOUR
+
+        self.action_bar_graph.set_data(bar_graph_data, colours=bar_graph_colours)
+
+        self.action_bar_graph.set_vertical_axis_label(
+            action_selection_strategy_info.action_value_description
+        )
+
+    def __save_linear_q_function_weights(self) -> None:
+        """Open a file dialog to save the weights of the currently selected 
+        LinearQFunction to a CSV file
+        """
+        mdp_name = self.mdp_combo_box.currentText()
+
+        if (mdp_name not in self.mdp_info_dict) or (
+            not self.mdp_info_dict[mdp_name].linear_q_func_info
+        ):
+            return
```

Comment:
nit: could be a helper function since it's duplicated in a bunch o' places in this file

Reply:
the attribute we access from the MdpInfo changes